### ISSUE TITLE
[no-explicit-any] Allow any inside extends section of template

### ISSUE THREAD

**Issue by Aliaksei-Martsinkevich** [typescript-eslint/typescript-eslint#817]
**Repro**
[playground](https://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAggdiAPAMTlCAPYE4BMDOUAFAHRkCGATgOb4BcU5CA2gLoCUUAvAHyMJ8uUVAFgAUKEhQAqnADWcAPYB3OCjSZseQqQo16UAK7ylqtp15GTKuIOFxx4ydGABGbrASJnigGZQAWXJgAAsSAFtyDB4AbidwFwAmD1kFG28Ev0DgsMjomKgAekKoXEUIfDgAcmAoZUVKOXEgA)
```TS
type Any<Fn extends (...args: any[]) => any> = Fn
type Unknown<Fn extends (...args: unknown[]) => unknown> = Fn

declare const sum: (a: number, b: number) => number;

type t1 = Any<typeof sum>; // (a: number, b: number) => number
type t2 = Unknown<typeof sum>; // doesn't work
```

typescript error:
```
Type '(a: number, b: number) => number' does not satisfy the constraint '(...args: unknown[]) => unknown'.
  Types of parameters 'a' and 'args' are incompatible.
    Type 'unknown' is not assignable to type 'number'.
```

**Expected Result**
I expect to have an option to allow `any` in `extends` section of type definition.
**Actual Result**
I don't have an option to enable `any` in templates
**Additional Info**
When I write code use generics with type restrictions, it's common case to fill generic arguments with `any`. In this case, type safety isn't violated, because it will not be `any` when it is instantiated. Quite often it is possible to use `never` instead, but it's not always the case.

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.13.0` |
| `@typescript-eslint/parser`        | `1.13.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.1.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.10.2` |

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md#ignorerestargs

```ts
/*
eslint @typescript-eslint/no-explicit-any: ['error', { ignoreRestArgs: true }]
*/
type Any<Fn extends (...args: any[]) => any> = Fn
type Unknown<Fn extends (...args: any[]) => unknown> = Fn
//                                ^^^ this is now allowed

declare const sum: (a: number, b: number) => number;

type t1 = Any<typeof sum>; // (a: number, b: number) => number
type t2 = Unknown<typeof sum>; // (a: number, b: number) => number
```

**Comment by oatkiller** [comment]
@bradzacher 
Wouldn't that also allow stuff like:
```ts
function takesAnything(...args: any[]) {
  for (const couldBeAnything of args) {
    // no type safety around `couldBeAnything` now
  }
}
```
In my case, I want to avoid bindings having an `any` type, but I still want to write code like:

```
/**
 * Obtain the return type of a function type
 */
type ReturnType<T extends (...args: any) => any> = T extends (...args: any) => infer R ? R : any;
```

**Comment by bradzacher** [comment]
(pedantic but `ReturnType` is [built into typescript](https://github.com/microsoft/TypeScript/blob/5ae286329f254a9b5fed1dd2d2208f76d9b7e011/src/lib/es5.d.ts#L1469))

---

Thinking about this more, and reading a bit more into generic types.

Ignore my previous comment. It's the wrong way to do this.

You shouldn't ever use any. This option is not required for defining types.
Instead, you can use `never` or `unknown`.

In the case of generic type constraints - you can use `never`:
```ts
type ReturnTypeT<T extends (...args: never[]) => any> = T extends (...args: any) => infer R ? R : any;

type MyFunc1 = () => string;
type MyFunc2 = (k: string) => number;
type MyFunc3 = (k: string, k2: number) => void;
type MyFunc4 = (...args: string[]) => boolean;

type RT1 = ReturnTypeT<MyFunc1> // typeof === string
type RT2 = ReturnTypeT<MyFunc2> // typeof === number
type RT3 = ReturnTypeT<MyFunc3> // typeof === void
type RT4 = ReturnTypeT<MyFunc4> // typeof === boolean
```
[repl](https://www.typescriptlang.org/play/#code/C4TwDgpgBAShwFcBOA7AKuCaA8aoQA9gIUATAZygAoA6OgQyQHNyAuKFCANwiQG0AugEooAXgB8UeihCTRUPIWJlKtBszZSZIiVACWKAGa9YUAPyn20kAG4AUHdCQoAWRAAxBCgDGARjHUOpLkwEgGTPZO0G6ePgBMAVQA1uwhYShMQRwIALYARryRmK4eXt4AzIkpUGnhADRQSXHsKLkFSFlcAPZ6pEXOMWUALIl0NIwsqaHhgll5XV0ANhDS9o7FMGj+8nCIqBiQOIM+vuLrzpsJO-DI6JhHpfFnUbBoldd7d4fYxxXPG2gRh9bgcsD9Ht4hmcgA)
